### ISSUE TITLE
NoSuchMethodError when using Spring Spanner starter

### ISSUE THREAD

**Issue by JoePacman** [spring-cloud/spring-cloud-gcp#1815]
Hi there,

Been trying to fix an error for the last couple of days in a microservice I'm building to interact with spanner in app engine standard (built in Spring and using cloud endpoints). The error I'm getting looks very similar to https://github.com/spring-cloud/spring-cloud-gcp/issues/1735.

The error looks like this in appengine:

```
com.google.api.server.spi.SystemService invokeServiceMethod: exception occurred while calling backend method (SystemService.java:373)
com.google.cloud.spanner.SpannerException: UNKNOWN: java.lang.NoSuchMethodError: io.opencensus.tags.TagContextBuilder.putPropagating(Lio/opencensus/tags/TagKey;Lio/opencensus/tags/TagValue;)Lio/opencensus/tags/TagContextBuilder;
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:155)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:101)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:82)
	at com.google.cloud.spanner.SessionPool$Waiter.take(SessionPool.java:827)
	at com.google.cloud.spanner.SessionPool$Waiter.access$4100(SessionPool.java:813)
	at com.google.cloud.spanner.SessionPool.getReadSession(SessionPool.java:1182)
	at com.google.cloud.spanner.DatabaseClientImpl.getReadSession(DatabaseClientImpl.java:52)
	at com.google.cloud.spanner.DatabaseClientImpl.singleUse(DatabaseClientImpl.java:104)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.doWithOrWithoutTransactionContext(SpannerTemplate.java:586)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.getReadContext(SpannerTemplate.java:129)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.performQuery(SpannerTemplate.java:440)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.executeQuery(SpannerTemplate.java:408)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.queryAndResolveChildren(SpannerTemplate.java:521)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.query(SpannerTemplate.java:212)
	at org.springframework.cloud.gcp.data.spanner.repository.query.SpannerStatementQueryExecutor.executeQuery(SpannerStatementQueryExecutor.java:85)
	at org.springframework.cloud.gcp.data.spanner.repository.query.PartTreeSpannerQuery.executeRawResult(PartTreeSpannerQuery.java:66)
	at org.springframework.cloud.gcp.data.spanner.repository.query.AbstractSpannerQuery.execute(AbstractSpannerQuery.java:64)
	at org.springframework.cloud.gcp.data.spanner.repository.query.PartTreeSpannerQuery.execute(PartTreeSpannerQuery.java:35)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:605)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy83.findByGCPCountryCodeAndBusinessCodeAndPortfolioStatusCodeAndContractTypeCodeAndClientReferenceAndAccountNumber(Unknown Source)
	at com.ri.service.PremiseServiceImpl.findByGCPCountryCodeAndBusinessCodeAndPortfolioStatusCodeAndContractTypeCodeAndClientReferenceAndAccountNumber(PremiseServiceImpl.java:30)
	at com.ri.controller.PremiseController.getPremiseSingleAccountNumber(PremiseController.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.api.server.spi.SystemService.invokeServiceMethod(SystemService.java:349)
	at com.google.api.server.spi.handlers.EndpointsMethodHandler$RestHandler.handle(EndpointsMethodHandler.java:127)
	at com.google.api.server.spi.handlers.EndpointsMethodHandler$RestHandler.handle(EndpointsMethodHandler.java:110)
	at com.google.api.server.spi.dispatcher.PathDispatcher.dispatch(PathDispatcher.java:50)
	at com.google.api.server.spi.EndpointsServlet.service(EndpointsServlet.java:80)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1772)
	at com.google.api.control.ControlFilter.doFilter(ControlFilter.java:331)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.api.control.ConfigFilter.doFilter(ConfigFilter.java:125)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:109)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:109)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:109)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:128)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.access$000(ErrorPageFilter.java:66)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:103)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:109)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:121)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:109)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:582)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at com.google.apphosting.runtime.jetty9.ParseBlobUploadHandler.handle(ParseBlobUploadHandler.java:119)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1182)
	at com.google.apphosting.runtime.jetty9.AppEngineWebAppContext.doHandle(AppEngineWebAppContext.java:187)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at com.google.apphosting.runtime.jetty9.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:293)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:539)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:333)
	at com.google.apphosting.runtime.jetty9.RpcConnection.handle(RpcConnection.java:213)
	at com.google.apphosting.runtime.jetty9.RpcConnector.serviceRequest(RpcConnector.java:81)
	at com.google.apphosting.runtime.jetty9.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:134)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:722)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:685)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:655)
	at com.google.apphosting.runtime.JavaRuntime$NullSandboxRequestRunnable.run(JavaRuntime.java:847)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:270)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.google.cloud.spanner.SpannerException: UNKNOWN: java.lang.NoSuchMethodError: io.opencensus.tags.TagContextBuilder.putPropagating(Lio/opencensus/tags/TagKey;Lio/opencensus/tags/TagValue;)Lio/opencensus/tags/TagContextBuilder;
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:155)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:101)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:82)
	at com.google.cloud.spanner.SessionPool$4.run(SessionPool.java:1536)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:295)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: com.google.cloud.spanner.SpannerException: UNKNOWN: java.lang.NoSuchMethodError: io.opencensus.tags.TagContextBuilder.putPropagating(Lio/opencensus/tags/TagKey;Lio/opencensus/tags/TagValue;)Lio/opencensus/tags/TagContextBuilder;
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:155)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:45)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:112)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get(GapicSpannerRpc.java:596)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.createSession(GapicSpannerRpc.java:475)
	at com.google.cloud.spanner.SpannerImpl.createSession(SpannerImpl.java:173)
	at com.google.cloud.spanner.SessionPool$4.run(SessionPool.java:1527)
	... 7 more
Caused by: java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: io.opencensus.tags.TagContextBuilder.putPropagating(Lio/opencensus/tags/TagKey;Lio/opencensus/tags/TagValue;)Lio/opencensus/tags/TagContextBuilder;
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:500)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:459)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get(GapicSpannerRpc.java:590)
	... 10 more
Caused by: java.lang.NoSuchMethodError: io.opencensus.tags.TagContextBuilder.putPropagating(Lio/opencensus/tags/TagKey;Lio/opencensus/tags/TagValue;)Lio/opencensus/tags/TagContextBuilder;
	at io.grpc.internal.CensusStatsModule$ClientCallTracer.<init>(CensusStatsModule.java:359)
	at io.grpc.internal.CensusStatsModule.newClientCallTracer(CensusStatsModule.java:146)
	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor.interceptCall(CensusStatsModule.java:689)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at com.google.api.gax.grpc.GrpcChannelUUIDInterceptor.interceptCall(GrpcChannelUUIDInterceptor.java:52)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at com.google.api.gax.grpc.GrpcHeaderInterceptor.interceptCall(GrpcHeaderInterceptor.java:80)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at com.google.api.gax.grpc.GrpcMetadataHandlerInterceptor.interceptCall(GrpcMetadataHandlerInterceptor.java:55)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor.interceptCall(SpannerErrorInterceptor.java:64)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at com.google.cloud.spanner.spi.v1.LoggingInterceptor.interceptCall(LoggingInterceptor.java:68)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at io.grpc.internal.ManagedChannelImpl.newCall(ManagedChannelImpl.java:814)
	at io.grpc.internal.ForwardingManagedChannel.newCall(ForwardingManagedChannel.java:63)
	at io.grpc.stub.MetadataUtils$HeaderAttachingClientInterceptor.interceptCall(MetadataUtils.java:74)
	at io.grpc.ClientInterceptors$InterceptorChannel.newCall(ClientInterceptors.java:156)
	at com.google.api.gax.grpc.GrpcClientCalls.newCall(GrpcClientCalls.java:90)
	at com.google.api.gax.grpc.GrpcDirectCallable.futureCall(GrpcDirectCallable.java:58)
	at com.google.api.gax.grpc.GrpcUnaryRequestParamCallable.futureCall(GrpcUnaryRequestParamCallable.java:65)
	at com.google.api.gax.grpc.GrpcExceptionCallable.futureCall(GrpcExceptionCallable.java:64)
	at com.google.api.gax.rpc.AttemptCallable.call(AttemptCallable.java:86)
	at com.google.api.gax.rpc.RetryingCallable.futureCall(RetryingCallable.java:63)
	at com.google.api.gax.rpc.RetryingCallable.futureCall(RetryingCallable.java:41)
	at com.google.api.gax.tracing.TracedUnaryCallable.futureCall(TracedUnaryCallable.java:75)
	at com.google.api.gax.rpc.UnaryCallable$1.futureCall(UnaryCallable.java:126)
	... 10 more

```

Following that thread I've tried upgrading the spring-cloud-gcp-starter-data-spanner to a milestone version  (1) and using the maven enforcer plugin (including explicitly loading in the latest version of opencensus) and making sure that there are no issues there (2)

Pom for (1):
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.1.6.RELEASE</version>
		<relativePath/> <!-- lookup parent from repository -->
	</parent>
	<groupId>com.api.csi</groupId>
	<artifactId>wf-spanner-service</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>war</packaging>
	<name>wf-spanner-service</name>
	<description>Workflows spanner microservice</description>

	<properties>
		<java.version>1.8</java.version>
		<appengine.maven.plugin>1.3.2</appengine.maven.plugin>
		<endpoints.project.id>#####REMOVING FROM THIS POST####</endpoints.project.id>
		<servlet.class.doc>com.google.api.server.spi.EndpointsServlet</servlet.class.doc>
	</properties>

	<repositories>
		<repository>
			<id>spring-snapshots</id>
			<name>Spring SnapShots</name>
			<url>https://repo.spring.io/libs-snapshot</url>
			<snapshots>
				<enabled>true</enabled>
			</snapshots>
		</repository>
	</repositories>

	<dependencies>

		<!-- Spring dependencies -->
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-gcp-starter-data-spanner</artifactId>
			<version>1.2.0.BUILD-SNAPSHOT</version>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<version>2.1.6.RELEASE</version>
			<scope>test</scope>
		</dependency>

		<!-- Lombok annotations -->
		<dependency>
			<groupId>org.projectlombok</groupId>
			<artifactId>lombok</artifactId>
			<version>1.18.8</version>
			<optional>true</optional>
		</dependency>

		<!-- Google - required dependencies -->
		<dependency>
			<groupId>com.google.endpoints</groupId>
			<artifactId>endpoints-management-control-appengine-all</artifactId>
			<version>1.0.11</version>
		</dependency>
		<dependency>
			<groupId>com.google.apis</groupId>
			<artifactId>google-api-services-cloudkms</artifactId>
			<version>v1-rev20190426-1.28.0</version>
		</dependency>
		<dependency>
			<groupId>com.google.appengine</groupId>
			<artifactId>appengine-api-1.0-sdk</artifactId>
			<version>1.9.75</version>
		</dependency>
		<dependency>
			<groupId>com.google.endpoints</groupId>
			<artifactId>endpoints-framework</artifactId>
			<version>2.2.1</version>
			<exclusions>
				<exclusion>
					<groupId>org.slf4j</groupId>
					<artifactId>slf4j-nop</artifactId>
				</exclusion>
			</exclusions>
		</dependency>

	</dependencies>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-gcp-dependencies</artifactId>
				<version>1.2.0.BUILD-SNAPSHOT</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
			<dependency>
				<groupId>com.google.cloud</groupId>
				<artifactId>google-cloud-bom</artifactId>
				<version>0.104.0-alpha</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<profiles>
		<profile>
			<id>dev-asia</id>
			<properties>
				<logging.level>DEBUG</logging.level>
			</properties>
			<build>
				<resources>
					<resource>
						<directory>src/main/resources/dev-asia</directory>
					</resource>
				</resources>
			</build>
		</profile>
	</profiles>

	<build>
		<outputDirectory>${project.build.directory}/${project.build.finalName}/WEB-INF/classes</outputDirectory>
		<plugins>

			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>

			<plugin>
				<groupId>com.google.cloud.tools</groupId>
				<artifactId>appengine-maven-plugin</artifactId>
				<version>2.0.0</version>
				<configuration>
					<projectId>${endpoints.project.id}</projectId>
					<version>GCLOUD_CONFIG</version>
				</configuration>
			</plugin>

			<plugin>
				<groupId>com.google.cloud.tools</groupId>
				<artifactId>endpoints-framework-maven-plugin</artifactId>
				<version>2.0.1</version>
				<configuration>
					<hostname>wf-spanner-service-dot-${endpoints.project.id}.appspot.com</hostname>
				</configuration>
			</plugin>

			<plugin>
				<artifactId>maven-war-plugin</artifactId>
				<version>3.2.3</version>
				<configuration>
					<failOnMissingWebXml>false</failOnMissingWebXml>
					<webResources>
						<resources>
							<directory>${basedir}/src/main/webapp/WEB-INF</directory>
							<filtering>true</filtering>
							<targetPath>WEB-INF</targetPath>
						</resources>
					</webResources>
				</configuration>
			</plugin>

		</plugins>
	</build>

</project>

```

Pom for (2):

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.1.6.RELEASE</version>
		<relativePath/> <!-- lookup parent from repository -->
	</parent>
	<groupId>com.api.csi</groupId>
	<artifactId>wf-spanner-service</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>war</packaging>
	<name>wf-spanner-service</name>
	<description>Workflows spanner microservice</description>

	<properties>
		<java.version>1.8</java.version>
		<spring-cloud.version>Greenwich.SR2</spring-cloud.version>
		<appengine.maven.plugin>1.3.2</appengine.maven.plugin>
		<endpoints.project.id>#####REMOVING FROM THIS POST####</endpoints.project.id>
		<servlet.class.doc>com.google.api.server.spi.EndpointsServlet</servlet.class.doc>
	</properties>

	<dependencies>

		<!-- First 2 sets of dependencies required due to debugging issues similar to https://github.com/spring-cloud/spring-cloud-gcp/issues/1735 -->
		<!-- Miscellaneous - explicitly declared dependencies -->
		<dependency>
			<groupId>io.opencensus</groupId>
			<artifactId>opencensus-api</artifactId>
			<version>0.23.0</version>
		</dependency>
		<dependency>
			<groupId>io.opencensus</groupId>
			<artifactId>opencensus-contrib-grpc-util</artifactId>
			<version>0.23.0</version>
		</dependency>
		<dependency>
			<groupId>io.grpc</groupId>
			<artifactId>grpc-context</artifactId>
			<version>1.21.0</version>
		</dependency>
		<dependency>
			<groupId>com.fasterxml.jackson.core</groupId>
			<artifactId>jackson-annotations</artifactId>
			<version>2.9.6</version>
		</dependency>


		<!-- Google - explicitly declared dependencies -->
		<dependency>
			<groupId>com.google.cloud</groupId>
			<artifactId>google-cloud-core-grpc</artifactId>
			<version>1.78.0</version>
		</dependency>
		<dependency>
			<groupId>com.google.guava</groupId>
			<artifactId>guava</artifactId>
			<version>28.0-jre</version>
		</dependency>
		<dependency>
			<groupId>com.google.cloud</groupId>
			<artifactId>google-cloud-logging</artifactId>
			<version>1.78.0</version>
		</dependency>
		<dependency>
			<groupId>com.google.http-client</groupId>
			<artifactId>google-http-client-jackson2</artifactId>
			<version>1.31.0</version>
		</dependency>

		<!-- Spring cloud dependencies version is set using <spring-cloud.version> -->
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-gcp-starter-data-spanner</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<version>2.1.6.RELEASE</version>
			<scope>test</scope>
		</dependency>

		<dependency>
			<groupId>org.projectlombok</groupId>
			<artifactId>lombok</artifactId>
			<version>1.18.8</version>
			<optional>true</optional>
		</dependency>

		<!-- Google - required depedencies -->
		<dependency>
			<groupId>com.google.endpoints</groupId>
			<artifactId>endpoints-management-control-appengine-all</artifactId>
			<version>1.0.11</version>
		</dependency>
		<dependency>
			<groupId>com.google.apis</groupId>
			<artifactId>google-api-services-cloudkms</artifactId>
			<version>v1-rev20190426-1.28.0</version>
		</dependency>
		<dependency>
			<groupId>com.google.appengine</groupId>
			<artifactId>appengine-api-1.0-sdk</artifactId>
			<version>1.9.75</version>
		</dependency>
		<dependency>
			<groupId>com.google.endpoints</groupId>
			<artifactId>endpoints-framework</artifactId>
			<version>2.2.1</version>
			<exclusions>
				<exclusion>
					<groupId>org.slf4j</groupId>
					<artifactId>slf4j-nop</artifactId>
				</exclusion>
			</exclusions>
		</dependency>

	</dependencies>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-dependencies</artifactId>
				<version>${spring-cloud.version}</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<profiles>
		<profile>
			<id>dev-asia</id>
			<properties>
				<logging.level>DEBUG</logging.level>
			</properties>
			<build>
				<resources>
					<resource>
						<directory>src/main/resources/dev-asia</directory>
					</resource>
				</resources>
			</build>
		</profile>
	</profiles>

	<build>
		<outputDirectory>${project.build.directory}/${project.build.finalName}/WEB-INF/classes</outputDirectory>
		<plugins>


			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-enforcer-plugin</artifactId>
				<version>1.4.1</version>
				<executions>
					<execution>
						<id>enforce</id>
						<goals>
							<goal>enforce</goal>
						</goals>
						<configuration>
							<rules>
								<requireUpperBoundDeps/>
							</rules>
						</configuration>
					</execution>
				</executions>
			</plugin>


			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>

			<plugin>
				<groupId>com.google.cloud.tools</groupId>
				<artifactId>appengine-maven-plugin</artifactId>
				<version>2.0.0</version>
				<configuration>
					<projectId>${endpoints.project.id}</projectId>
					<version>GCLOUD_CONFIG</version>
				</configuration>
			</plugin>

			<plugin>
				<groupId>com.google.cloud.tools</groupId>
				<artifactId>endpoints-framework-maven-plugin</artifactId>
				<version>2.0.1</version>
				<configuration>
					<hostname>wf-spanner-service-dot-${endpoints.project.id}.appspot.com</hostname>
				</configuration>
			</plugin>

			<plugin>
				<artifactId>maven-war-plugin</artifactId>
				<version>3.2.3</version>
				<configuration>
					<failOnMissingWebXml>false</failOnMissingWebXml>
					<webResources>
						<resources>
							<directory>${basedir}/src/main/webapp/WEB-INF</directory>
							<filtering>true</filtering>
							<targetPath>WEB-INF</targetPath>
						</resources>
					</webResources>
				</configuration>
			</plugin>

		</plugins>
	</build>

</project>

```

Unfortunately I'm still having no luck. Any ideas? Apologies if I've missed something really obvious, I'm quite new to Spring and GCP! Thanks! :)

**Comment by elefeint** [comment]
It looks like a mismatch between the OpenCensus version that Spanner uses and the one that you have in the project.

What happens when you run this command?
```
mvn dependency:tree -Dincludes=io.opencensus
```

**Comment by JoePacman** [comment]
It shows what looks like the same version of opencensus-api:0.21.0:

```[INFO] com.api.csi:wf-spanner-service:war:0.0.1-SNAPSHOT
[INFO] \- org.springframework.cloud:spring-cloud-gcp-starter-data-spanner:jar:1.2.0.BUILD-SNAPSHOT:compile
[INFO]    \- org.springframework.cloud:spring-cloud-gcp-data-spanner:jar:1.2.0.BUILD-SNAPSHOT:compile
[INFO]       \- com.google.cloud:google-cloud-spanner:jar:1.30.0:compile
[INFO]          +- com.google.cloud:google-cloud-core-grpc:jar:1.85.0:compile
[INFO]          |  \- com.google.api:gax-grpc:jar:1.47.1:compile
[INFO]          |     \- io.grpc:grpc-netty-shaded:jar:1.21.0:compile
[INFO]          |        \- io.grpc:grpc-core:jar:1.21.0:compile (version selected from constraint [1.21.0,1.21.0])
[INFO]          |           \- io.opencensus:opencensus-contrib-grpc-metrics:jar:0.21.0:compile
[INFO]          \- io.opencensus:opencensus-contrib-grpc-util:jar:0.21.0:compile
[INFO]             \- io.opencensus:opencensus-api:jar:0.21.0:compile

```

**Comment by meltsufin** [comment]
Looks like another diamond dependency problem involving opencensus and gRPC.
@suztomo @elharo

**Comment by suztomo** [comment]
I found artifact `com.google.endpoints:endpoints-management-control-appengine-all:1.0.11` contains old `io.opencensus.tags.TagContextBuilder`.

@JoePacman Would you try setting priority to `opencensus-api-0.23.0.jar`? [Java 8 Runtime Environment: Class loader JAR ordering](
https://cloud.google.com/appengine/docs/standard/java/runtime#Java%208_Class_loader_JAR_ordering)

**Comment by JoePacman** [comment]
@suztomo many thanks for your response. That has fixed the opencensus exception that was been thrown :)

Now my deployment gets a timeout waiting on Spanner Sessions, but I imagine that is something related to how I've integrated cloud endpoints with Spring and I need to do some further investigation.

**Comment by suztomo** [comment]
Glad to hear it fixed the problem. Feel free to create a new issue if you need any help.

**Comment by meltsufin** [comment]
This should have been resolved with the addition of the libraries-bom.
